File: _Programmers-고득점-Kit/02_스택_큐/PG-Lv2_올바른 괄호.java
import java.util.*;

class Solution {
    boolean solution(String s) {
        
        Stack<Character> stack = new Stack<>();
        for (char p : s.toCharArray()) {
            if (p == '(') {
                stack.push(p);
            } else if (!stack.empty()) {
                stack.pop();
            } else {
                return false;
            }
        }

//         if (stack.empty()) {
//             return true;
//         } else {
//             return false;
//         }
        
        return (stack.empty() ? true : false);
        
    }
}

File: _Programmers-고득점-Kit/02_스택_큐/PG-Lv2_올바른 괄호.py
# [PG] 12909 - 올바른 괄호 (Lv2)
# https://school.programmers.co.kr/learn/courses/30/lessons/12909

def solution(s):
    stack = []

    for i in s:
        if i == '(':
            stack.append(i)
        else:  # -- i == ')'
            if stack and stack[-1] == '(':
                stack.pop()
            else:
                stack.append(i)

    return False if stack else True


def solution2(s):
    stack = []
    for p in s:
        if p == "(":
            stack.append(p)
        elif stack:
            stack.pop()
        else:
            return False

    return len(stack) == 0


s = "(())()"
assert True == solution(s)
